[libc++] basic_ios<wchar_t> cannot store fill character WCHAR_MAX

xingxue-ibm opened the pull request:
`libcxx std::basic_ios` uses `WEOF` to indicate the `fill` value is uninitialized. On some platforms (e.g AIX and zOS in 64-bit mode) `wchar_t` is 4 bytes `unsigned` and `wint_t` is also 4 bytes which means `WEOF` cannot be distinguished from `WCHAR_MAX` by `std::char_traits<wchar_t>::eq_int_type()`, meaning this valid character value cannot be stored on affected platforms (as the implementation triggers reinitialization to `widen(’ ’)`).

This patch introduces a new helper class `_FillHelper` uses a boolean variable to indicate whether the fill character has been initialized, which is used by default in libcxx ABI version 2. The patch does not affect ABI version 1 except for targets AIX in 32- and 64-bit and z/OS in 64-bit (so that the layout of the implementation is compatible with the current IBM system provided libc++)

This is a continuation of Phabricator patch [D124555](https://reviews.llvm.org/D124555). This patch uses a modified version of the [approach](https://reviews.llvm.org/D124555#3566746) suggested by @ldionne .

xingxue-ibm committed:
Use lower max threads to reduce the testing time on AIX.

xingxue-ibm committed:
Changes for cases where wint_t and wchar_t have the same width. The changes are compatible with the current IBM provided libc++.

xingxue-ibm commented:
@perry-ca

perry-ca commented on libcxx/include/ios:
```diff
@@ -521,6 +522,31 @@ inline _LIBCPP_HIDE_FROM_ABI void ios_base::exceptions(iostate __iostate) {
 }
 
+template <class _Traits>
+// Attribute 'packed' is used to keep the layout compatible with the previous
```
Can you expand this comment to explain this doesn't give 100% compat.  It will eliminate the tail padding and keep the member after this one in the right offset.  However, being packed the alignment of this struct will be 1 instead of the alignment of int_type.  If this struct was added as a member where the offset wasn't already a multiple of the alignment of int_type then the offset would not be preserved.

I don't think you can take the int_type/bool pair and put them in a struct that will preserve the size and alignment of the original pair.  You can either get the alignment correct or the size correct and not both.

xingxue-ibm committed:
Addressed comments.
- update a comment

daltenty commented on libcxx/include/ios:
```diff
@@ -590,7 +616,24 @@ protected:
   basic_ostream<char_type, traits_type>* __tie_;
-  mutable int_type __fill_;
+
+#if defined(_AIX) || (defined(__MVS__) && defined(__64BIT__))
+// AIX and 64-bit MVS must use _OptionalFill for ABI backward compatibility.
+  using _FillType = _OptionalFill<_Traits>;
+#else
+#if defined(_WIN32)
+  static const bool _OptOutForABICompat = true;
+#else
+  static const bool _OptOutForABICompat = false;
+#endif
+
+  using _FillType = _If<
+      sizeof(char_type) >= sizeof(int_type) && !_OptOutForABICompat,
+      _OptionalFill<_Traits>,
+      _SentinelValueFill<_Traits>
+  >;
+#endif
```
I feel like this platform ABI logic belongs in `include/__configuration/abi.h` (see Louis' recent refactor in https://github.com/llvm/llvm-project/commit/23e1ed65c2c3eb1f80f7eeb4897ec843492c5451), with flags like `_LIBCXX_IOS_MAY_USE_OPTIONAL_FILL` and `_LIBCXX_IOS_FORCE_OPTIONAL_FILL`

In ABI version 2, `_LIBCXX_IOS_MAY_USE_OPTIONAL_FILL` could simply be on, even on AIX and 64-bit MVS. In ABI version 1, we enable `_LIBCXX_IOS_FORCE_OPTIONAL_FILL` for AIX  and 64-bit MVS and leave every thing else alone (that way we don't need to special case Windows like we do here)

Then this becomes:
```suggestion
#if defined(_LIBCXX_IOS_FORCE_OPTIONAL_FILL)
// AIX and 64-bit MVS must use _OptionalFill for ABI backward compatibility.
  using _FillType = _OptionalFill<_Traits>;
#elif defined(_LIBCXX_IOS_MAY_USE_OPTIONAL_FILL)
  using _FillType = _If<
      sizeof(char_type) >= sizeof(int_type),
      _OptionalFill<_Traits>,
      _SentinelValueFill<_Traits>
  >;
#else
  using _FillType = _SentinelValueFill<_Traits>;
#endif
```

xingxue-ibm replied:
Thanks very much for the suggestion, @daltenty! This makes it fits nicely in the platform ABI logic.

xingxue-ibm committed:
Addressed comments.
-define and use macros _LIBCXX_IOS_MAY_USE_OPTIONAL_FILL and _LIBCXX_IOS_FORCE_OPTIONAL_FILL.

xingxue-ibm committed:
Add _LIBCPP_HIDE_FROM_ABI.

xingxue-ibm committed:
Removed changes in unrelated test case.

xingxue-ibm committed:
Add description of ABI change in ReleaseNote.

daltenty commented on libcxx/test/std/input.output/iostream.format/std.manip/setfill_eof.pass.cpp:
```diff
@@ -0,0 +1,31 @@
+// Test that weof as a wchar_t value can be set as the fill character.
+
+// UNSUPPORTED: no-wide-characters
+// REQUIRES: target=powerpc{{(64)?}}-ibm-aix || target=s390x-ibm-zos
```
I think we can expect this test to pass on all targets which don't have to opt out for ABI compat (i.e. everything but windows in v1 ABI), so this should be something like:
```suggestion
// XFAIL: target=={{.+}}-windows-{{.+}} && stdlib=libc++ && libcpp-abi-version=1
```

xingxue-ibm replied:
Good point! Changed as suggested, thanks!

xingxue-ibm committed:
Address comment.
- expect the test case to pass except windows in abi version 1.

daltenty commented on libcxx/include/ios:
```diff
@@ -605,7 +643,7 @@ template <class _CharT, class _Traits>
   ios_base::init(__sb);
   __tie_  = nullptr;
-  __fill_ = traits_type::eof();
+  __fill_ = widen(' ');
```
There is some history to why widen is not called here:
https://github.com/llvm/llvm-project/commit/3320c058bf4e7b827d8546b31881ff043c3a44df

perry-ca replied:
It would be worth adding this as a comment.  It's so easy to think calling widen() here is the better thing to do.

xingxue-ibm replied:
Removed `__fill_ = widen(' ');` because the original statement `__fill_ = traits_type::eof();` sets the fill value to `traits_type::eof()` to indicate it is uninitialized. This is now accomplished by constructors, `_SentinelValueFill() : __fill_val_(_Traits::eof()) { }` and ` _FillHelper() : __set_(false) { }`.

daltenty replied:
> This is now accomplished by constructors

basic_ios has a bunch of rather interesting initialization requirements according to the standard (related to why it has an init function at all), and the constructor isn't really supposed to initialize in some cases.

I think to preserve that, and the original flow of the code, you should add an Init function to the both fill types in place of the constructor and call it here.

xingxue-ibm replied:
Thanks for the insight! Changed to use an init function as suggested.

xingxue-ibm committed:
Use helper class _FillHelper by default in ABI version 2.

xingxue-ibm committed:
Remove unnecessary include of __type_traits/conditional.h.

daltenty commented on libcxx/include/__configuration/abi.h:
```diff
@@ -91,6 +91,16 @@
+// and WCHAR_MAX. New helper class _FillHelper uses a boolean variable to indicate
+// whether the fill value has been initialized so that a fill value WEOF set
+// by the user won't be treated as indicating the fill value is uninitialized.
+// Undefining macro _LIBCXX_IOS_USE_FILL_HELPER to keep the ABI verson 1
+// behavior if needed.
```
I think this comment is a bit confusing (there is no undef here, nor is it need at the moment) and probably redundant (of course you can not def the macro to opt out, that's generally the same for every macro in this config file)
```suggestion
```

xingxue-ibm replied:
Removed the sentence.

xingxue-ibm committed:
Addressed comments.
- add note that the change only affect ABI version 2
- remove "undefining _LIBCXX_IOS_USE_FILL_HELPER to keep the ABI verson 1 behavior"
- add init() functions for class _FillHelper and _SentinelValueFill

xingxue-ibm commented on libcxx/docs/ReleaseNotes/19.rst:
```diff
@@ -146,6 +146,12 @@ ABI Affecting Changes
   ``random_device`` could throw a ``system_error`` with this value. It now
   throws ``ENOMSG``.
 
+- libcxx ``std::basic_ios`` uses ``WEOF`` to indicate that the fill value is uninitialized. However, on platforms
```
Changed as suggested, thanks!

xingxue-ibm committed:
Rename init() to __init().

daltenty commented on libcxx/include/ios:
```diff
@@ -590,7 +615,13 @@ protected:
   basic_ostream<char_type, traits_type>* __tie_;
-  mutable int_type __fill_;
+
+#if defined(_LIBCXX_IOS_USE_FILL_HELPER)
+  using _FillType = _FillHelper<_Traits>;
+#else
+  using _FillType = _SentinelValueFill<_Traits>;
+#endif
+  mutable _FillType __fill_;
```
minor nit: I'd suggest `traits_type` instead of `_Traits` as the parameter here for consistency which the style of the rest of the class definition:
```suggestion
#if defined(_LIBCXX_IOS_USE_FILL_HELPER)
  using _FillType = _FillHelper<traits_type>;
#else
  using _FillType = _SentinelValueFill<traits_type>;
#endif
  mutable _FillType __fill_;
```

xingxue-ibm replied:
Good catch, thanks!

daltenty reviewed (approved):
LGTM, thanks! (with a minor style nit)

xingxue-ibm committed:
Addressed comments.
- Use traits_type instead of _Traits to be consistent.

daltenty commented:
Gentle ping @reviewers-libcxx

ldionne commented on libcxx/test/std/input.output/iostream.format/std.manip/setfill_eof.pass.cpp:
```diff
@@ -0,0 +1,31 @@
+// Test that weof as a wchar_t value can be set as the fill character.
+
+// UNSUPPORTED: no-wide-characters
+// REQUIRES: target=powerpc{{(64)?}}-ibm-aix || target=s390x-ibm-zos || libcpp-abi-version=2
```
On what platforms does this actually fail? Can we simply get rid of this `REQUIRES`? I'd like to see what fails in our CI if we simply remove this line.

xingxue-ibm replied:
@ldionne The test case fails the CI on various targets after `REQUIRES` is removed. This is because without it, it goes through `libcpp-abi-version=1` path and therefore, uses class `_SentinelValueFill` for these targets. Setting the `__fill_val_` to `WEOF` causes the `os.fill()` call in `assert` to re-initialize `__fill_val` with `widen(` `)` and uses it as the return value and therefore, fails the assert.

As per discussion with @daltenty, we think it is wrong to set `WEOF` for the test because it is not a valid wide character. We should use `WCHAR_MAX` instead. By setting `WCHAR_MAX`, we can test the implementation on AIX and zOS where `WEOF` is the same as `WCHAR_MAX` as well as on other targets without `REQUIRES`. The test case would fail on other targets if `WEOF` is also the same as `WCHAR_MAX`.

ldionne reviewed (changes requested):
This looks sensible, but I have some feedback.

xingxue-ibm committed:
Update libcxx/include/__configuration/abi.h

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

xingxue-ibm committed:
Update libcxx/include/__configuration/abi.h

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

xingxue-ibm commented on libcxx/include/__configuration/abi.h:
```diff
@@ -91,6 +91,14 @@
+// and WCHAR_MAX. New helper class _FillHelper uses a boolean variable to indicate
+// whether the fill value has been initialized so that a fill value WEOF set
+// by the user won't be treated as indicating the fill value is uninitialized.
+#  define _LIBCXX_IOS_USE_FILL_HELPER
```
I like it, thanks!

xingxue-ibm committed:
Update libcxx/include/__configuration/abi.h

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

xingxue-ibm committed:
Update libcxx/include/ios

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

xingxue-ibm committed:
Addressed comments.
- set cmake variable LIBCXX_ABI_DEFINES with the ABI macro for AIX and zOS 64-bit
- remove the note for this change from the release note.
- rename function name __fill() to __get()
- remove REQUIRES from the test case.

xingxue-ibm committed:
Remove the unused macro definition for AIX/zOS in abi.h.

xingxue-ibm committed:
Merge branch 'main' into ios-fill

xingxue-ibm committed:
setfill() WCHAR_MAX instead of WEOF for the test.

xingxue-ibm committed:
Fixed format.

xingxue-ibm commented:
> LGTM, but CI seems to be failing right now.

Thanks so much @ldionne. Yeah, the CI failed on the following targets where `WEOF` is the same as `WCHAR_MAX`. I've updated the test case to `XFAIL` those targets with libcpp ABI version 1.
```
target=aarch64-linux-gnu
target=armv8l-linux-gnueabihf
target=armv7l-linux-gnueabihf
target=x86_64-pc-windows-msvc
target=x86_64-w64-windows-gnu
target=i686-w64-windows-gnu
```

xingxue-ibm committed:
XFAIL targets where WEOF is the same as WCHAR_MAX in libcpp ABI version 1.

daltenty commented on libcxx/test/std/input.output/iostream.format/std.manip/setfill_wchar_max.pass.cpp:
```diff
@@ -0,0 +1,38 @@
+// version 2 implementation fixes the problem.
+
+// XFAIL: target={{.*}}-windows{{.*}} && libcpp-abi-version=1
+// XFAIL: target=armv{{71|81}}-linux-gnueabihf && libcpp-abi-version=1
```
@xingxue-ibm I think there's a typo here (i.e. `l` for little endian not `1`) :
```suggestion
// XFAIL: target=armv{{7|8}}l-linux-gnueabihf && libcpp-abi-version=1
```

xingxue-ibm replied:
Ah, good catch, thanks!

xingxue-ibm committed:
Update libcxx/test/std/input.output/iostream.format/std.manip/setfill_wchar_max.pass.cpp

Co-authored-by: David Tenty <daltenty.dev@gmail.com>

xingxue-ibm commented:
Thanks very much, @ldionne and @daltenty for your guidance and help to get the patch in the shape as it is now!

vitalybuka commented:
Looks like this breaks:
https://lab.llvm.org/buildbot/#/builders/85/builds/705
https://lab.llvm.org/buildbot/#/builders/24/builds/587

daltenty commented:
> Looks like this breaks: https://lab.llvm.org/buildbot/#/builders/85/builds/705 https://lab.llvm.org/buildbot/#/builders/24/builds/587

This test is already XFAIL'd for aarch64, but it seems the XFAIL in the test did not match on these bots because they have a four component triple in their test run. 

I've update the XFAIL condition with ad154281230d83ee551e12d5be48bb956ef47ed3, which should hopefully resolve the breakage.

xingxue-ibm commented:
> > Looks like this breaks: https://lab.llvm.org/buildbot/#/builders/85/builds/705 https://lab.llvm.org/buildbot/#/builders/24/builds/587
> 
> This test is already XFAIL'd for aarch64, but it seems the XFAIL in the test did not match on these bots because they have a four component triple in their test run.
> 
> I've update the XFAIL condition with [ad15428](https://github.com/llvm/llvm-project/commit/ad154281230d83ee551e12d5be48bb956ef47ed3), which should hopefully resolve the breakage.

Thanks. @daltenty!

zeroomega commented:
Hi, we saw a series of weird test failures on our mac stage2 builder after this patch was landed. And after hours of bisecting and we confirmed it was caused by this patch. The error message we saw on our bot:

```
FAIL: LLVM :: tools/llvm-debuginfo-analyzer/COFF/01-coff-print-basic-details.test (48643 of 54789)
******************** TEST 'LLVM :: tools/llvm-debuginfo-analyzer/COFF/01-coff-print-basic-details.test' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 21: /Volumes/Work/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/llvm-debuginfo-analyzer --attribute=level,format                          --output-sort=offset                          --print=scopes,symbols,types,lines,instructions                          /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-debuginfo-analyzer/COFF/Inputs/test-codeview-clang.o 2>&1 |  /Volumes/Work/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/FileCheck --strict-whitespace -check-prefix=ONE /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-debuginfo-analyzer/COFF/01-coff-print-basic-details.test
+ /Volumes/Work/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/FileCheck --strict-whitespace -check-prefix=ONE /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-debuginfo-analyzer/COFF/01-coff-print-basic-details.test
+ /Volumes/Work/s/w/ir/x/w/llvm_build/tools/clang/stage2-bins/bin/llvm-debuginfo-analyzer --attribute=level,format --output-sort=offset --print=scopes,symbols,types,lines,instructions /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-debuginfo-analyzer/COFF/Inputs/test-codeview-clang.o
/Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-debuginfo-analyzer/COFF/01-coff-print-basic-details.test:43:13: error: ONE-NEXT: expected string not found in input
; ONE-NEXT: [004]     5             {Line}
            ^
<stdin>:11:61: note: scanning from here
[004]                   {Variable} 'CONSTANT' -> 'const int'
                                                            ^
<stdin>:12:1: note: possible intended match here
[004] ����5             {Line}
^

Input file: <stdin>
Check file: /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-debuginfo-analyzer/COFF/01-coff-print-basic-details.test

-dump-input=help explains the following input dump.

Input was:
<<<<<<
           .
           .
           .
           6: [002]               {Function} extern not_inlined 'foo' -> 'int' 
           7: [003]                 {Parameter} 'ParamPtr' -> '* const int' 
           8: [003]                 {Parameter} 'ParamUnsigned' -> 'unsigned' 
           9: [003]                 {Parameter} 'ParamBool' -> 'bool' 
          10: [003]                 {Block} 
          11: [004]                   {Variable} 'CONSTANT' -> 'const int' 
next:43'0                                                                 X error: no match found
          12: [004] ����5             {Line} 
next:43'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:43'1     ?                               possible intended match
          13: [004]                   {Code} 'movl	$0x7, 0x4(%rsp)' 
next:43'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          14: [004] ����6             {Line} 
next:43'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          15: [004]                   {Code} 'movl	$0x7, 0x1c(%rsp)' 
next:43'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          16: [004]                   {Code} 'jmp	0x8' 
next:43'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          17: [003]                 {TypeAlias} 'INTEGER' -> 'int' 
next:43'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           .
           .
           .
>>>>>>

--

********************
```

You can see that the clang emitted some non-readable characters in this unit test. It doesn't happen in this way before this PR was landed.

Our mac stage2 toolchain is linked against to the stage1 toolchain's libcxx, that probably explained why the stage1 builder is green but the test failed on stage2, since the stage1 toolchain used host toolchain's libcxx, which doesn't have this PR. The issue only happens on Mac, looks like Windows and Linux, are unaffected. 
Example bot run with this issue:  https://ci.chromium.org/ui/p/fuchsia/builders/prod/clang-mac-x64/b8742128214337927377/overview

Could you take a look please? If it takes a while to fix, could you revert this change and land it later?

ldionne commented:
@xingxue-ibm The most likely thing here is that this patch changed the size, alignment, trailing padding or some triviality property of `basic_ios` when the ABI flag is disabled. Maybe the `_LIBCPP_PACKED` attribute is to blame? Either way, if we can't figure it out quickly, we should probably revert this because we are close to the LLVM 19 release and we want the tree to be as stable as possible.

Also, once we figure it out, we should add a test for that.

daltenty commented:
An initial thing note is this is this buildbot seems to be explicit building the version 2 ABI, not the stable one:
https://github.com/llvm/llvm-project/blob/1f6f97e2b64a9c4c51712d32b53e07d2a20e0cc0/clang/cmake/caches/Fuchsia.cmake#L129

I'll continue investigation

daltenty commented:
@zeroomega we’ve taken a look at the buildbot config and outputs.

From what we can see your bots stage 1 use a clang with a custom v2 libc++ from cpid:
```
 ‘-DCMAKE_C_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang’,
 ‘-DCMAKE_CXX_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang++‘,
 ‘-DCMAKE_ASM_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang’,
…
 ‘-DCMAKE_SHARED_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a’,
 ‘-DCMAKE_MODULE_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a’,
 ‘-DCMAKE_EXE_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a’,
```
https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8742128214337927377/+/u/clang/configure/execution_details

And then we have some more cmake flags for stage2 which seem to be set to link the cpid library, not the just built one in the build tree:
```
 ‘-DSTAGE2_CROSS_TOOLCHAIN_FLAGS_NATIVE=-DCMAKE_C_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang;-DCMAKE_CXX_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang++;-DCMAKE_OSX_SYSROOT=/Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk;-DCMAKE_SHARED_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a;-DCMAKE_MODULE_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a;-DCMAKE_EXE_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a’,
```
Our current theory based on this is your stage2 built clang is loading the cpid shipped libc++ instead of the just-built libc++ from stage 1.

That could explain the problems your are seeing, since we changed the offset of the fill character in the layout of basic_ios in the unstable v2 ABI with this patch.

Are you able to try running the tests with a custom `DYLD_LIBRARY_PATH` pointed at the builds `lib` directory (the one containing the just-built libc++) to verify if that resolves the test failure? It would be good to check the rpath, etc on the stage2 built clang as well to confirm what version of libc++ is actually being loading.

(You can reach me as `daltenty` on LLVM discord if you’d like to discuss further. I've been trying to replicate the full bot config locally on an x64 mac but there's a lot of moving parts in that bot.)

zeroomega commented:
> @zeroomega we’ve taken a look at the buildbot config and outputs.
> 
> From what we can see your bots stage 1 use a clang with a custom v2 libc++ from cpid:
> 
> ```
>  ‘-DCMAKE_C_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang’,
>  ‘-DCMAKE_CXX_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang++‘,
>  ‘-DCMAKE_ASM_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang’,
> …
>  ‘-DCMAKE_SHARED_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a’,
>  ‘-DCMAKE_MODULE_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a’,
>  ‘-DCMAKE_EXE_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a’,
> ```
> 
> https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8742128214337927377/+/u/clang/configure/execution_details
> 
> And then we have some more cmake flags for stage2 which seem to be set to link the cpid library, not the just built one in the build tree:
> 
> ```
>  ‘-DSTAGE2_CROSS_TOOLCHAIN_FLAGS_NATIVE=-DCMAKE_C_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang;-DCMAKE_CXX_COMPILER=/Volumes/Work/s/w/ir/x/w/cipd/bin/clang++;-DCMAKE_OSX_SYSROOT=/Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk;-DCMAKE_SHARED_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a;-DCMAKE_MODULE_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a;-DCMAKE_EXE_LINKER_FLAGS=-nostdlib++ /Volumes/Work/s/w/ir/x/w/cipd/lib/libc++.a’,
> ```
> 
> Our current theory based on this is your stage2 built clang is loading the cpid shipped libc++ instead of the just-built libc++ from stage 1.
> 
> That could explain the problems your are seeing, since we changed the offset of the fill character in the layout of basic_ios in the unstable v2 ABI with this patch.
> 
> Are you able to try running the tests with a custom `DYLD_LIBRARY_PATH` pointed at the builds `lib` directory (the one containing the just-built libc++) to verify if that resolves the test failure? It would be good to check the rpath, etc on the stage2 built clang as well to confirm what version of libc++ is actually being loaded.
> 
> (You can reach me as `daltenty` on LLVM discord if you’d like to discuss further. I've been trying to replicate the full bot config locally on an x64 mac but there's a lot of moving parts in that bot.)

I don't think your theory is correct, as the CROSS_NATIVE flag is only going to be effective when doing cross compilation. We do have a cross compile mac-arm64 builder that why those flags were there (we should clean them up when doing native host build). In addition, when I was bisecting locally, I omit the CROSS_NATIVE flags. The flag set i used are:

```
/Users/haowei/SRC/llvm-prebuilts/cmake/mac-amd64/bin/cmake \
  -S \
  /Users/haowei/SRC/llvm-project/llvm \
  -GNinja \
  -DCMAKE_MAKE_PROGRAM=/Users/haowei/SRC/llvm-prebuilts/ninja/mac-amd64/ninja \
  -DCMAKE_INSTALL_PREFIX= \
  -DCMAKE_C_COMPILER=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/clang \
  -DCMAKE_CXX_COMPILER=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/clang++ \
  -DCMAKE_ASM_COMPILER=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/clang \
  -DCLANG_REPOSITORY_STRING=https://llvm.googlesource.com/llvm-project \
  -DCMAKE_LIBTOOL=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/llvm-libtool-darwin \
  -DCMAKE_LIPO=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/llvm-lipo \
  -DCMAKE_SYSROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk \
  -DLLVM_ENABLE_ZLIB=FORCE_ON \
  -DZLIB_INCLUDE_DIR=/Users/haowei/SRC/llvm-prebuilts/reproduce/zlib_install_target/include \
  -DZLIB_LIBRARY=/Users/haowei/SRC/llvm-prebuilts/reproduce/zlib_install_target/lib/libz.a \
  -DLLVM_ENABLE_ZSTD=FORCE_ON \
  -Dzstd_DIR=/Users/haowei/SRC/llvm-prebuilts/reproduce/zstd_install/lib/cmake/zstd \
  -DLLVM_ENABLE_LIBXML2=FORCE_ON \
  -DLibXml2_ROOT=/Users/haowei/SRC/llvm-prebuilts/reproduce/libxml2_install_target/lib/cmake/libxml2-2.9.10 \
  -DLLVM_ENABLE_CURL=FORCE_ON \
  -DCURL_ROOT=/Users/haowei/SRC/llvm-prebuilts/reproduce/curl_install/lib/cmake/CURL \
  -DOpenSSL_ROOT=/Users/haowei/SRC/llvm-prebuilts/reproduce/boringssl_install/lib/cmake/OpenSSL \
  -DLLVM_ENABLE_HTTPLIB=FORCE_ON \
  -Dhttplib_ROOT=/Users/haowei/SRC/llvm-prebuilts/reproduce/cpp_httplib_install/lib/cmake/httplib \
  -DCMAKE_FIND_PACKAGE_PREFER_CONFIG=ON \
  "-DCMAKE_SHARED_LINKER_FLAGS=-nostdlib++ /Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a" \
  "-DCMAKE_MODULE_LINKER_FLAGS=-nostdlib++ /Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a" \
  "-DCMAKE_EXE_LINKER_FLAGS=-nostdlib++ /Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a" \
  -DSTAGE2_CMAKE_LINKER=/usr/bin/ld \
  -DSTAGE2_LINUX_aarch64-unknown-linux-gnu_SYSROOT=/Users/haowei/SRC/llvm-prebuilts/sysroot/linux \
  -DSTAGE2_LINUX_armv7-unknown-linux-gnueabihf_SYSROOT=/Users/haowei/SRC/llvm-prebuilts/sysroot/linux \
  -DSTAGE2_LINUX_i386-unknown-linux-gnu_SYSROOT=/Users/haowei/SRC/llvm-prebuilts/sysroot/linux \
  -DSTAGE2_LINUX_riscv64-unknown-linux-gnu_SYSROOT=/Users/haowei/SRC/llvm-prebuilts/sysroot/focal \
  -DSTAGE2_LINUX_x86_64-unknown-linux-gnu_SYSROOT=/Users/haowei/SRC/llvm-prebuilts/sysroot/linux \
  -DSTAGE2_FUCHSIA_SDK=/Users/haowei/SRC/llvm-prebuilts/sdk \
  "-DSTAGE2_LLVM_LIT_ARGS=--resultdb-output=r.j -v" \
  -DSTAGE2_LLVM_ENABLE_LTO=OFF \
  -DSTAGE2_LLVM_RAEVICT_MODEL_PATH=none \
  -DCURSES_INCLUDE_DIRS=/Users/haowei/SRC/llvm-prebuilts/reproduce/ncurses_install/include\;/Users/haowei/SRC/llvm-prebuilts/reproduce/ncurses_install/include/ncurses \
  -DCURSES_LIBRARIES=/Users/haowei/SRC/llvm-prebuilts/reproduce/ncurses_install/lib/libncurses.a \
  -DPANEL_LIBRARIES=/Users/haowei/SRC/llvm-prebuilts/reproduce/ncurses_install/lib/libpanel.a \
  -DLLVM_ENABLE_TERMINFO=FORCE_ON \
  -DTerminfo_LIBRARIES=/Users/haowei/SRC/llvm-prebuilts/reproduce/ncurses_install/lib/libncurses.a \
  -DLLVM_ENABLE_LIBEDIT=FORCE_ON \
  -DLibEdit_INCLUDE_DIRS=/Users/haowei/SRC/llvm-prebuilts/reproduce/libedit_install/include \
  -DLibEdit_LIBRARIES=/Users/haowei/SRC/llvm-prebuilts/reproduce/libedit_install/lib/libedit.a\;/Users/haowei/SRC/llvm-prebuilts/reproduce/ncurses_install/lib/libncurses.a \
  -C ../clang/cmake/caches/Fuchsia.cmake
```

So I don't think the issue is stage2 clang is loading stage1 libcxx. In addition, we always do static linking, I am not sure if `DYLD_LIBRARY_PATH` is going to be effective in this case but I will have a try when I am back to my workstation.

daltenty commented:
> So I don't think the issue is stage2 clang is loading stage1 libcxx. In addition, we always do static linking, I am not sure if `DYLD_LIBRARY_PATH` is going to be effective in this case but I will have a try when I am back to my workstation. 

Thanks for the additional information. That's a little mysterious then, the issue your seeing certainly seems to resemble a layout mismatch on `basic_ios`, but the only obvious way to end up with such a case is if the definition in the library being used and the definition in the headers disagree (i.e. using a version before the change in one place and the one after in the other).

I'll continue to try to reproduce locally with the information you've provided as well.

zeroomega commented:
@daltenty 

I further remove unnecessary cmake flags and it still produce the same error.

```
/Users/haowei/SRC/llvm-prebuilts/cmake/mac-amd64/bin/cmake \
  -S \
  /Users/haowei/SRC/llvm-project/llvm \
  -GNinja \
  -DCMAKE_MAKE_PROGRAM=/Users/haowei/SRC/llvm-prebuilts/ninja/mac-amd64/ninja \
  -DCMAKE_INSTALL_PREFIX= \
  -DCMAKE_C_COMPILER=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/clang \
  -DCMAKE_CXX_COMPILER=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/clang++ \
  -DCMAKE_ASM_COMPILER=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/clang \
  -DCLANG_REPOSITORY_STRING=https://llvm.googlesource.com/llvm-project \
  -DCMAKE_LIBTOOL=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/llvm-libtool-darwin \
  -DCMAKE_LIPO=/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/bin/llvm-lipo \
  -DCMAKE_SYSROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk \
  -DCMAKE_FIND_PACKAGE_PREFER_CONFIG=ON \
  "-DCMAKE_SHARED_LINKER_FLAGS=-nostdlib++ /Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a" \
  "-DCMAKE_MODULE_LINKER_FLAGS=-nostdlib++ /Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a" \
  "-DCMAKE_EXE_LINKER_FLAGS=-nostdlib++ /Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a" \
  "-DSTAGE2_LLVM_LIT_ARGS=--resultdb-output=r.j -v" \
  -DSTAGE2_LLVM_ENABLE_LTO=OFF \
  -DSTAGE2_LLVM_RAEVICT_MODEL_PATH=none \
  -DLLVM_ENABLE_LIBEDIT=FORCE_ON \
  -C ../clang/cmake/caches/Fuchsia.cmake
```

You can get our host clang toolchain from https://chrome-infra-packages.appspot.com/p/fuchsia/third_party/clang/mac-amd64/+/integration and you just need cmake and ninja and xcode to reproduce the build. 

I will try to see if I can get a more concrete confirmation that the libc++ for the stage2 build is coming from stage1 toolchain build.

zeroomega commented:
@daltenty Thanks for looking into this.  At this point I am 90% confident that the test failure on `01-coff-print-basic-details.test` is an issue on our side. 
I did a local build and rerun the stage2 build of `llvm-debuginfo-analyzer` (which is what was failing in the unit test) with `-v` and noticed that the stage 1's libc++ linker flag (`/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a`) was incorrectly passed through:

```
_  stage2-bins git:(bad) _ ninja llvm-debuginfo-analyzer -v
[1/1] : && /Users/haowei/SRC/llvm-project/build_stage2/./bin/clang++ --sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -stdlib=libc++ -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffile-prefix-map=/Users/haowei/SRC/llvm-project/build_stage2/tools/clang/stage2-bins=../../../../ -ffile-prefix-map=/Users/haowei/SRC/llvm-project/= -no-canonical-prefixes -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -mmacosx-version-min=10.13 -Wl,-search_paths_first -Wl,-headerpad_max_install_names -nostdlib++ /Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a -stdlib=libc++ -static-libstdc++ -fuse-ld=lld -Wl,--color-diagnostics -Wl,-lto_library -Wl,/Users/haowei/SRC/llvm-project/build_stage2/./lib/libLTO.dylib    -Wl,-dead_strip tools/llvm-debuginfo-analyzer/CMakeFiles/llvm-debuginfo-analyzer.dir/llvm-debuginfo-analyzer.cpp.o tools/llvm-debuginfo-analyzer/CMakeFiles/llvm-debuginfo-analyzer.dir/Options.cpp.o -o bin/llvm-debuginfo-analyzer  -Wl,-rpath,@loader_path/../lib  lib/libLLVMX86Desc.a  lib/libLLVMARMDesc.a  lib/libLLVMAArch64Desc.a  lib/libLLVMRISCVDesc.a  lib/libLLVMX86Disassembler.a  lib/libLLVMARMDisassembler.a  lib/libLLVMAArch64Disassembler.a  lib/libLLVMRISCVDisassembler.a  lib/libLLVMX86Info.a  lib/libLLVMARMInfo.a  lib/libLLVMAArch64Info.a  lib/libLLVMRISCVInfo.a  lib/libLLVMBinaryFormat.a  lib/libLLVMDebugInfoCodeView.a  lib/libLLVMDebugInfoLogicalView.a  lib/libLLVMDebugInfoPDB.a  lib/libLLVMMC.a  lib/libLLVMMCDisassembler.a  lib/libLLVMObject.a  lib/libLLVMSupport.a  lib/libLLVMARMDesc.a  lib/libLLVMARMInfo.a  lib/libLLVMARMUtils.a  lib/libLLVMAArch64Desc.a  lib/libLLVMCodeGenTypes.a  lib/libLLVMAArch64Info.a  lib/libLLVMAArch64Utils.a  lib/libLLVMRISCVDesc.a  lib/libLLVMRISCVInfo.a  lib/libLLVMMCDisassembler.a  lib/libLLVMDebugInfoMSF.a  lib/libLLVMDebugInfoDWARF.a  lib/libLLVMObject.a  lib/libLLVMIRReader.a  lib/libLLVMBitReader.a  lib/libLLVMAsmParser.a  lib/libLLVMCore.a  lib/libLLVMRemarks.a  lib/libLLVMBitstreamReader.a  lib/libLLVMMCParser.a  lib/libLLVMMC.a  lib/libLLVMDebugInfoCodeView.a  lib/libLLVMTextAPI.a  lib/libLLVMBinaryFormat.a  lib/libLLVMTargetParser.a  lib/libLLVMSupport.a  lib/libLLVMDemangle.a  -lm && :
clang++: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
clang++: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
clang++: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]

```

So the stage2 clang, when building this binary was linked to the host's libcxx. I will try to manually correct it to the stage2 libcxx to see if the failure disappear.

daltenty commented:
> @daltenty Thanks for looking into this. At this point I am 90% confident that the test failure on `01-coff-print-basic-details.test` is an issue on our side. I did a local build and rerun the stage2 build of `llvm-debuginfo-analyzer` (which is what was failing in the unit test) with `-v` and noticed that the stage 1's libc++ linker flag (`/Users/haowei/SRC/llvm-prebuilts/clang/mac-amd64/lib/libc++.a`) was incorrectly passed through:
> 
> ...
> 
> So the stage2 clang, when building this binary was linked to the host's libcxx. I will try to manually correct it to the stage2 libcxx to see if the failure disappear.

No problem, thanks for confirming (my local build is still running). That would definitely explain the issues you are seeing, since the two libraries disagree on the layout. 
(I haven't seen a lot of buildbots using this `-DSTAGE2_` mechanism so potentially there could be some CMake bugs lurking in that area)
